### ISSUE TITLE
Running M4C_captioner on sample images

### ISSUE THREAD

**Issue by freeIsa** [facebookresearch/mmf#663]
## ❓ Questions and Help
Hi there, I am starting to look into the task of image captioning and wanted to try out some pretrained models on a bunch of images I have. 
Inspired by `mmf_hm_example.ipynb` notebook, I tried to load the M4C_captioner model with these instructions:
```
from mmf.models.m4c_captioner import M4CCaptioner
model = M4CCaptioner.from_pretrained("m4c_captioner.textcaps.defaults")
```

but after downloading model weights the execution failed with the following error:
```
AttributeError: 'NoneType' object has no attribute '_select_impl'
	full_key: 
	reference_type=None
	object_type=None
```
(raised in class OmegaConf)

Is there a different way to load this model and run it on my custom images? Are there any tutorials around using pretrained MMF models for generating image captions I can refer to?
Thanks a lot!

**Comment by ytsheng** [comment]
In order to run image captioning on your custom images, you'd need to extract the image features from your custom images first. If you haven't done that, then you can use the scripts for it: https://github.com/facebookresearch/mmf/tree/master/tools/scripts/features 

I am currently working on a tutorial for image feature extraction that I will publish and link here early next week. (November 2nd, 2020 week). 

Thanks for using mmf! 😃

**Comment by ytsheng** [comment]
https://github.com/facebookresearch/mmf/pull/673/files Here is the tutorial code as promised. :) Let us know how else we can help.

**Comment by freeIsa** [comment]
Hi there,
first of all thank you for the tutorial!
I was about to try it out but unfortunately the installation of `vqa_maskrcnn-benchmark` on my machine fails because of incompatible packages.
As `vqa_maskrcnn-benchmark` appears to have been deprecated in favor of `Detectron2`, I was wondering whether it'd be possible to extract features for MMF using Detectron2 instead...

**Comment by freeIsa** [comment]
Hello again,
after several attempts I am still unable to install `vqa_maskrcnn-benchmark` and to make `extract_features_vmb` work without failing because of version conflicts at some point.
Can someone please provide a working configuration (e.g. pytorch, torch-vision version number)?
I am working on a Linux Server with python 3.8 and CUDA 10.2 (though I also tried with a conda environment forced to python 3.6), using conda environments. Any help would be greatly appreciated!

**Comment by revantteotia** [comment]
@freeIsa Are you able to run M4C_captioner on sample images? I have extracted both object and ocr frcn features for my sample images but I am not able to figure out which config files to update so that it uses my custom extracted features.
Thanks!

**Comment by freeIsa** [comment]
> @freeIsa Are you able to run M4C_captioner on sample images? I have extracted both object and ocr frcn features for my sample images but I am not able to figure out which config files to update so that it uses my custom extracted features.
> Thanks!

@revantteotia sadly, not yet. I still couldn't find a way to install the vqa_maskrcnn-benchmark needed to extract features on my machine. From my understanding, after obtaining image features one should implement a new Dataset and DatasetBuilder class to add to MMF registry so that the new dataset can be passed to the model (https://mmf.sh/docs/tutorials/dataset). Good luck!

**Comment by apsdehal** [comment]
@freeIsa  To install vqa-maskrcnn-benchmark, usually you need to install MMF first and then VQA-maskrcnn-benchmark. Did you try doing that?

**Comment by freeIsa** [comment]
@apsdehal I tried many things and I tried that too, but I keep finding conflicts between versions of python, CUDA, pytorch etc. I couldn't find any working configuration, alas!

**Comment by tranvinhhung99** [comment]
https://github.com/facebookresearch/mmf/blob/5c75d690ebbc0359a7e0b7719c0a8d4c996fd22e/mmf/models/m4c.py#L333

Hi, I also meet the same issue. I have found the about line is the cause of the error. Or at least I have found a way to load weight. I will try to see if I could infer in a later day.

In short, the existing config (which store in registry["config"]) for downloading model being modified by this line -> Change `config` to `config_2`.

To be noted further, I also edit config file of M4C (which is in `$HOME/.cache/torch/mmf/data/models/m4c_captioner.textcaps.defaults`) to change all absolute path to some path on my computer.

**Comment by shuaichenchang** [comment]
@ytsheng Hi, thank you for making the tutorial! The link about how to install MMF in maskrcnn_benchmark env seems internal, can you make it public? Thank you! The link is https://www.internalfb.com/intern/staticdocs/mmf/docs/getting_started/installation#install-from-source-recommended

**Comment by tranvinhhung99** [comment]
Hi @ytsheng,

Do the feature extractor (Faster RCNN)  with `default` of M4C Captioner use ResNet101 or ResNet152 as the backbone?

Thank you.

**Comment by apsdehal** [comment]
@tranvinhhung99 it is ResNext152 trained on Visual Genome with attribute loss. 
@shuaichenchang The doc is here: https://mmf.sh/docs/tutorials/image_feature_extraction

**Comment by tranvinhhung99** [comment]
@apsdehal how about the OCR Feature FRCNN, as I believe they came from the same model?

Correct me if I am wrong, but AFAIK the code provided in https://github.com/ronghanghu/vqa-maskrcnn-benchmark-m4c/tree/bbox_feat_extract/maskrcnn_benchmark only support ResNext101.

Thank you a lots.

**Comment by shuaichenchang** [comment]
@apsdehal Thanks a lot for the reply! The document contains how to install maskrcnn_benchmark in a separate env, but I think to get the feature extractor run I need to install mmf and maskrcnn_benchmark in one env.  This section (https://mmf.sh/docs/tutorials/image_feature_extraction/#prerequisites-setup) referred to an internal document (https://www.internalfb.com/intern/staticdocs/mmf/docs/getting_started/installation#install-from-source-recommended) about the setup. I am wondering how to get maskrcnn_benchmark and mmf work in the same env.

**Comment by jc138691** [comment]
Dear mmf-team, I am trying out your example how to: Generate prediction file on the validation set, via
    # https://mmf.sh/docs/projects/m4c_captioner/
    # mmf_predict datasets=textcaps \
    #     model=m4c_captioner \
    #     config=projects/m4c_captioner/configs/m4c_captioner/textcaps/defaults.yaml \
    #     env.save_dir=./save/m4c_captioner/defaults \
    #     run_type=val \
    #     checkpoint.resume_zoo=m4c_captioner.textcaps.defaults

And getting the following error:

WARNING 2021-04-01T17:30:33 | mmf: Key textcaps_num_final_outputs is not present in registry, returning default value of   File "/Users/jc138691/dev/comp/mmf_v210330/mmf/trainers/mmf_trainer.py", line 43, in load
    super().load()
  File "/Users/jc138691/dev/comp/mmf_v210330/mmf/trainers/base_trainer.py", line 30, in load
    self.load_model()
  File "/Users/jc138691/dev/comp/mmf_v210330/mmf/trainers/mmf_trainer.py", line 94, in load_model
    self.model = build_model(attributes)
  File "/Users/jc138691/dev/comp/mmf_v210330/mmf/utils/build.py", line 94, in build_model
    model.build()
  File "/Users/jc138691/dev/comp/mmf_v210330/mmf/models/m4c.py", line 45, in build
    self._build_output()
  File "/Users/jc138691/dev/comp/mmf_v210330/mmf/models/m4c.py", line 157, in _build_output
    num_choices -= self.config.classifier.ocr_max_num
TypeError: unsupported operand type(s) for -=: 'NoneType' and 'int'

Any suggestions on how to resolve it? Thanks, Dmitry

**Comment by apsdehal** [comment]
@jc138691 Please open up a new issue.
@shuaichenchang We will fix that link. It is actually pointing to https://mmf.sh/docs#install-from-source-recommended.

**Comment by nmalboubi** [comment]
@revantteotia were you able to figure out what to do after extracting OCR and image features? Any luck with generating a caption?

**Comment by revantteotia** [comment]
> @revantteotia were you able to figure out what to do after extracting OCR and image features? Any luck with generating a caption?

No luck with generating captions for custom images. However, I wanted to use the encoder part of the M4C architecture for image-text retrieval, so I reimplemented the model and trained it on my custom dataset using the extracted OCR and image features.
I found it easier to reimplement the m4c model than to use the mmf framework.

**Comment by nmalboubi** [comment]
@apsdehal I'm able to extract image features using: https://github.com/facebookresearch/mmf/blob/master/tools/scripts/features/extract_features_vmb.py

And I'm able to extract any words/text in the image through another OCR algorithm which returns each word in a picture, in the form of a list.

Can you shed some more detail on how we need to combine these for inference using the M4C_Captioner? And can you also link to the latest yml file/pytorch model for the pretrained M4C Captioner model?

**Comment by apsdehal** [comment]
Assigning to @ronghanghu who can explain this is in more detail.

**Comment by nmalboubi** [comment]
@ronghanghu would you have any updates to share?

**Comment by ronghanghu** [comment]
@nmalboubi Hi, a possible way to run on a custom image dataset is to build them as imdbs and extra their image and OCR features, through the following steps.

Step 1: build imdbs for your dataset following the format of TextCaps training and validation imbs (will be automatically downloaded when running M4C-Captioner on the TextCaps dataset) in `$MMF_DATA_DIR/datasets/textcaps/defaults/annotations/imdb_train.npy` and `$MMF_DATA_DIR/datasets/textcaps/defaults/annotations/imdb_val_filtered_by_image_id.npy` (where `$MMF_DATA_DIR` is typically `~/.cache/torch/mmf/data` in your machine). These imdbs contain OCR results (under `ocr_tokens` key) and normalized bounding boxes (i.e. in the range of `[0,1]`) of each detected objects (under `obj_normalized_boxes` key) and OCR tokens (under `ocr_normalized_boxes` key), and the boxes are in the format of [xmin, ymin, xmax, ymax]. You can put your new imdbs under `data/datasets/textcaps/defaults/annotations/` with different file names.

Step 2: install a custom maskrcnn-benchmark for feature extraction from https://github.com/ronghanghu/vqa-maskrcnn-benchmark-m4c, which allows extracting bounding box features from arbitrary input boxes.

Step 3: extra the image features with https://github.com/facebookresearch/mmf/blob/master/tools/scripts/features/extract_features_vmb.py and turn the feature folder into an lmdb using https://github.com/facebookresearch/mmf/blob/master/tools/scripts/features/lmdb_conversion.py. You can put the new lmdb under `$MMF_DATA_DIR/datasets/textvqa/defaults/features/`.

Step 4: extra the OCR boxes features over the previously built imdbs in Step 1 using  https://github.com/facebookresearch/mmf/blob/master/projects/m4c/scripts/extract_ocr_frcn_feature.py and turn the feature folder into an lmdb using https://github.com/facebookresearch/mmf/blob/master/tools/scripts/features/lmdb_conversion.py. You can put the new lmdb under `$MMF_DATA_DIR/datasets/textvqa/defaults/features/`.

Step 5: finally, create a new config file that points to the new imdbs and features. You can modify https://github.com/facebookresearch/mmf/blob/master/mmf/configs/datasets/textcaps/defaults.yaml to contain the follows 

```​​
    annotations:
      train:
      - <your new train imdb>
      val:
      - <your new val imdb>  # only one sample per image_id
      test:
      - <your new test imdb>  # only one sample per image_id
    features:
      train:
      - <your new train image feature lmdb>,<your new train ocr feature lmdb>
      val:
      - <your new val image feature lmdb>,<your new val ocr feature lmdb>
      test:
      - <your new test image feature lmdb>,<your new test ocr feature lmdb>
```
(see https://github.com/facebookresearch/mmf/blob/master/mmf/configs/datasets/textcaps/defaults.yaml#L23-L29 and ​​https://github.com/facebookresearch/mmf/blob/master/projects/m4c_captioner/configs/m4c_captioner/textcaps/defaults.yaml#L14-L20

Step 6: one can train and evaluate M4C-Captioner with the updated config file following https://mmf.sh/docs/projects/m4c_captioner

Hope this helps!

**Comment by nmalboubi** [comment]
@ronghanghu appreciate the reply on this! I understand steps 2-5, but not sure I understand step 1. How would I create new imdbs for the annotations of my images? And if I'm not training the model, do I need to do this? I'm looking to build an inference job to do the prediction (not looking to do any training). Ideally, I can feed in one picture and receive the caption, so I'm thinking steps 2-5 would be relevant here. 

It seems like I would constantly be needing to update the yaml file based on the image I feed the model. I'd like to deploy this model to an endpoint, so not sure if I would be able to readjust the yaml file for each new image that is fed into the model (could be incorrect about this I'm not entirely certain).

**Comment by ronghanghu** [comment]
Hi @nmalboubi, for batch prediction, one can create an imdb following the format in `$MMF_DATA_DIR/datasets/textcaps/defaults/annotations/imdb_val_filtered_by_image_id.npy` (where `$MMF_DATA_DIR` is typically `~/.cache/torch/mmf/data` in your machine). This is basically a numpy file that contains a list of annotations, each annotation is an image and relevant information (such as object detection results, OCR tokens and boxes). If only inference is needed, one can build an imdb only for the validation set. The purpose of imdb is to put all relevant information together so that the TextCaps dataset class can load them into a sample_list and pass it to the model.

If the goal is to get an interactive system that takes one image and outputs a caption, one can manually feed the relevant data into a sample_list following the dataset loading in https://github.com/facebookresearch/mmf/blob/master/mmf/datasets/builders/textvqa/dataset.py#L85-L117. In this case, one can first use `build_model` in https://github.com/facebookresearch/mmf/blob/master/mmf/utils/build.py#L73 to construct a TextCaps model (from an existing config file like https://github.com/facebookresearch/mmf/blob/master/mmf/configs/datasets/textcaps/defaults.yaml) and then call `model.forward(sample_list)` on the manually constructed `sample_list` to get outputs from the model.

**Comment by nmalboubi** [comment]
Thanks @ronghanghu ! Appreciate the quick reply on this. Couple of clarifying questions: I guess how is the imdb file created in steps 3 and 4 different from that in Step 1? I've already completed steps 3 and 4, just not sure how it relates to Step 1? 

 For the interactive system, would I also need to complete step 1?

**Comment by ronghanghu** [comment]
Hi @nmalboubi, for Step 4 (extracting OCR box features), one can use the imdb built in Step 1 as input `--imdb_file` to https://github.com/facebookresearch/mmf/blob/master/projects/m4c/scripts/extract_ocr_frcn_feature.py, which will supply the OCR bounding boxes in https://github.com/facebookresearch/mmf/blob/master/projects/m4c/scripts/extract_ocr_frcn_feature.py#L181-L182. For Step 3 (extracting image features), the imdb is not specifically needed and as long as the Faster R-CNN features on each image are extracted and turned into an lmdb file, it should be good.

**Comment by nmalboubi** [comment]
@ronghanghu apologies for all the questions. I really appreciate you making the time to answer so quickly. I think I'm getting very close! Although, I think there's a miscommunication. I don't understand what we're building in Step 1, nor is it clear how I can build it. If I'm feeding a new image into the model, how am I supposed to be providing OCR results and features in Step 1? Isn't that what steps 3 and 4 do (extracting the OCR and features)? I have been able to build step 3, ignoring all of step 1. It sounds like Step 4 depends on Step 1, but it's not clear how I need to build that necessary file for my test image in Step 1 (again, I'm only using this model for inference). Would you please explain?

**Comment by nmalboubi** [comment]
@ronghanghu apologies for all the questions. I really appreciate you making the time to answer so quickly. I think I'm getting very close! Clarifying question: Would you please explain how to build the imdb file for Step 1 (if it's necessary for doing inference only on a new, unseen image)? Is it as simple as converting the image to a npy array? Thanks again for your assistance through this.

**Comment by ronghanghu** [comment]
Hi @nmalboubi, the imdb is a numpy file containing the information of all the images you would like to run upon for the TextCaps dataset to load all the image information. (Since you are only doing the inference case, you can just build an imdb for the test set, follow the TextCaps test set imdb `$MMF_DATA_DIR/datasets/textcaps/defaults/annotations/imdb_test_filtered_by_image_id.npy` (where `$MMF_DATA_DIR` is typically ~/.cache/torch/mmf/data in your machine).

Building this imdb numpy file requires putting relevant information into a Python list of dictionaries, and save this list to a numpy `.npy` file.

The TextCaps imdb for inference is numpy array of image information (Python dictionaries). An example list element (for a specific image) is the following (it does not contain the image files or feature vectors, but only paths to them):
```
{'caption_id': 300000000,  # you can an integer you like here
 'image_height': 1024,  # image sizes
 'image_width': 679,
 'feature_path': 'd7495e8d036dd20b.npy', # the paths to the feature file name
 'image_id': 'd7495e8d036dd20b',                        
 'image_name': 'd7495e8d036dd20b',                      
 'image_path': 'test/d7495e8d036dd20b.jpg',
 'ocr_tokens': ['adida', 'dilas', 'add', 'KHY'], # a list of OCR tokens for this image
 'ocr_normalized_boxes': array([ # normalized bounding boxes of these OCR tokens in [xmin, ymin, xmax, ymax] format
       [0.27412063, 0.32063147, 0.41068   , 0.3515165 ],
       [0.52118105, 0.33020005, 0.62584215, 0.36027625],
       [0.58363354, 0.542395  , 0.6346887 , 0.56811285],
       [0.27133596, 0.88131714, 0.35654247, 0.9217752 ]], dtype=float32),
 'obj_normalized_boxes': array([ # normalized bounding boxes of detected objects in [xmin, ymin, xmax, ymax] format
       [0.15631638, 0.18397395, 0.43572935, 0.4972998 ],
       [0.5273716 , 0.4913926 , 0.77885246, 0.62227875],
       [0.21730828, 0.71529853, 0.31667167, 0.886147  ],
       [0.49155214, 0.0517989 , 0.6399429 , 0.1169441 ],
       [0.18002696, 0.68084425, 0.5856489 , 0.8627959 ],
       [0.23358001, 0.05816906, 0.38508427, 0.14679454],
       [0.22738099, 0.880806  , 0.36854836, 0.93160856],
       [0.1914896 , 0.4882285 , 0.40171307, 0.58157724],
       [0.08597744, 0.05537106, 0.491714  , 0.8271646 ],
       [0.        , 0.6905568 , 0.99875   , 1.0070618 ],
       [0.30508482, 0.68489283, 0.4310713 , 0.8202953 ],
       [0.49959823, 0.20210488, 0.7835941 , 0.5218848 ],
       [0.23126411, 0.6137074 , 0.508905  , 0.89829296],
       [0.5307318 , 0.70756173, 0.65870243, 0.9254602 ],
       [0.41074938, 0.14257857, 0.8676123 , 0.8152167 ],
       [0.49102476, 0.9104151 , 0.68421966, 0.98049724],
       [0.        , 0.47040117, 0.86166996, 0.9510974 ],
       [0.23421526, 0.8102367 , 0.30867893, 0.90282404],
       [0.23401731, 0.7305831 , 0.73547775, 0.9092536 ],
       [0.13515711, 0.20879017, 0.23015748, 0.48773828],
       [0.1526573 , 0.8715652 , 0.39123794, 0.94313633],
       [0.18575278, 0.4593628 , 0.48043284, 0.66189545],
       [0.23299535, 0.68483657, 0.46707752, 0.7891162 ],
       [0.2130655 , 0.88550264, 0.3765384 , 0.9657019 ],
       [0.40701637, 0.03408793, 0.80604905, 0.43814766],
       [0.16304894, 0.5035279 , 0.2603272 , 0.5697723 ],
       [0.13774747, 0.60451734, 0.59104294, 0.98753273],
       [0.5279094 , 0.91394657, 0.81504774, 0.9825984 ],
       [0.51743686, 0.6449014 , 0.7250832 , 0.93932486],
       [0.09972609, 0.2723047 , 0.48859844, 0.46622097],
       [0.58891994, 0.9303476 , 0.6258184 , 0.9698728 ],
       [0.5028002 , 0.49701148, 0.78874475, 0.7343463 ],
       [0.1382074 , 0.8563756 , 0.74664605, 0.9827516 ],
       [0.19740562, 0.7580816 , 0.3652102 , 0.8988646 ],
       [0.8110651 , 0.23010021, 0.99255157, 0.6178901 ],
       [0.28872842, 0.6114304 , 0.78349924, 1.0070618 ],                                                              
       [0.19131124, 0.6997895 , 0.34514165, 0.84653115],                                                              
       [0.52425754, 0.1971732 , 0.80024767, 0.6121116 ],
       [0.5539563 , 0.15507999, 0.99875   , 0.6124007 ],
       [0.30828533, 0.08710695, 0.7383106 , 0.9296322 ],                                                              
       [0.39424455, 0.1170062 , 0.89092207, 0.5791726 ],
       [0.1252242 , 0.57916147, 0.99875   , 1.0070618 ],
       [0.6916775 , 0.50640863, 0.77824515, 0.5684655 ],
       [0.18313637, 0.7162816 , 0.4957798 , 0.9371931 ],
       [0.68354636, 0.8163399 , 0.72994107, 0.8430757 ],
       [0.16751847, 0.04875814, 0.4326949 , 0.3782226 ],
       [0.21533054, 0.7825018 , 0.32188645, 0.9175777 ],
       [0.49156928, 0.07804417, 0.64510596, 0.19676487],
       [0.21309198, 0.1935392 , 0.49539185, 0.58481044],
       [0.14721146, 0.        , 0.549387  , 0.4609366 ],
       [0.65856093, 0.9194021 , 0.74318606, 0.9705187 ],
       [0.21657163, 0.69441384, 0.40716803, 0.8412767 ],
       [0.8559029 , 0.27820927, 0.99875   , 0.41148087],
       [0.10820269, 0.10590867, 0.73424906, 0.5733099 ],
       [0.10958922, 0.47400478, 0.6523081 , 0.6431564 ],
       [0.2271672 , 0.05883767, 0.38668352, 0.19890423],
       [0.07256512, 0.32718456, 0.99875   , 0.92317545],
       [0.        , 0.        , 0.99875   , 0.3166512 ],
       [0.5586036 , 0.9230993 , 0.6459006 , 0.9746683 ],
       [0.25584918, 0.14730388, 0.37673545, 0.19104607],
       [0.5460971 , 0.7064676 , 0.792265  , 0.9132635 ],
       [0.1664785 , 0.00860521, 0.86212456, 0.44997957],
       [0.5018792 , 0.16145588, 0.8221138 , 0.42646688],
       [0.84421337, 0.26772514, 0.99875   , 0.59949315],
       [0.14557002, 0.13594401, 0.99875   , 0.8355856 ],
       [0.66357046, 0.186317  , 0.8590603 , 0.54564255],
       [0.54985934, 0.13152458, 0.5861344 , 0.15671879],
       [0.5719622 , 0.5330194 , 0.6407394 , 0.5746809 ],
       [0.48886633, 0.5563707 , 0.69855624, 0.9490296 ],
       [0.2534583 , 0.81335706, 0.29218492, 0.8393204 ],
       [0.5818796 , 0.5361416 , 0.62772197, 0.56360227],
       [0.12097656, 0.1768072 , 0.35258323, 0.62736017],
       [0.23655754, 0.18197295, 0.36747068, 0.22660698],
       [0.24862915, 0.39115325, 0.9670706 , 0.9643968 ],
       [0.1505656 , 0.48510975, 0.24500893, 0.5518546 ],
       [0.10464617, 0.17284264, 0.18548584, 0.46829054],
       [0.27405187, 0.12788579, 0.31002185, 0.15366574],
       [0.23923   , 0.5096811 , 0.5053857 , 0.88538265],
       [0.5072632 , 0.04476408, 0.99875   , 0.45413896],
       [0.        , 0.15011178, 0.47735006, 0.61224025],
       [0.58084184, 0.9381186 , 0.6163445 , 0.9608468 ],
       [0.4603183 , 0.9046656 , 0.77669907, 1.0070618 ],
       [0.5371324 , 0.57325023, 0.7343431 , 0.68320495],
       [0.2344232 , 0.71796536, 0.3075435 , 0.8221941 ],
       [0.18237537, 0.68612725, 0.44471568, 0.8808831 ],
       [0.35273775, 0.50417614, 0.4926527 , 0.59523505],
       [0.03428354, 0.        , 0.3428737 , 0.92371464],
       [0.11557673, 0.20542938, 0.18409035, 0.5142846 ],
       [0.        , 0.        , 0.40602118, 0.4882544 ],
       [0.4618013 , 0.08984942, 0.727173  , 0.47665155],
       [0.16371201, 0.57828265, 0.379983  , 0.92341506],
       [0.26196694, 0.5155139 , 0.30116904, 0.55568445],
       [0.40234023, 0.5062478 , 0.8961067 , 0.64629245],
       [0.6422885 , 0.74138975, 0.7435461 , 0.95448303],
       [0.38106993, 0.44532546, 0.99875   , 0.9523586 ],
       [0.3134799 , 0.677623  , 0.4301091 , 0.7827492 ],
       [0.4582453 , 0.06136325, 0.67009264, 0.15966602],
       [0.20072517, 0.6953184 , 0.3071111 , 0.85939676],
       [0.61512256, 0.10411632, 0.639956  , 0.14418834],
       [0.09105231, 0.18220092, 0.2357754 , 0.5896701 ]], dtype=float32),
 'set_name': 'test'
}
```

**Comment by Pxtri2156** [comment]
@ronghanghu  I am trying to extract OCR boxes features on the TextVQA dataset. I used annotations of the authors and `extract_ocr_frcn_feature.py` to extract features. But I same this error below: 
```
File "projects/m4c/scripts/extract_ocr_frcn_feature.py", line 199, in <module>  
    main()
  File "projects/m4c/scripts/extract_ocr_frcn_feature.py", line 189, in main
    detection_model, image_path, input_boxes=ocr_boxes
  File "projects/m4c/scripts/extract_ocr_frcn_feature.py", line 105, in extract_features
    output = detection_model(current_img_list, input_boxes=input_boxes)
  File "/root/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
TypeError: forward() got an unexpected keyword argument 'input_boxes'
```
Can you help me? 
Thank you very much.

**Comment by ronghanghu** [comment]
Hi @Pxtri2156, for this feature extraction, it is necessary to install `vqa-maskrcnn-benchmark` from https://github.com/ronghanghu/vqa-maskrcnn-benchmark-m4c to extract features from a specific bounding box.

**Comment by soon22** [comment]
@nmalboubi 
Hi, 
may you share what version of Python, CUDA and PyTorch you use to install `maskrcnn-benchmark` and `mmf` ?

I want to extract image features with:

https://github.com/facebookresearch/mmf/blob/master/tools/scripts/features/extract_features_vmb.py

But unable to get the libraries to work together
